libphal/phal_sbe.C

```diff
@@ -272,7 +271,92 @@ sbeError_t captureFFDC(struct pdbg_target *target)
 	// create ffdc file
 	tmpfile_t ffdcFile(bufPtr.getData(), ffdcLen);
 	return sbeError_t(exception::SBE_CMD_FAILED, ffdcFile.getFd(),
-			  ffdcFile.getPath().c_str());
+			  ffdcFile.getPath());
+}
+
+sbeError_t capturePOZFFDC(struct pdbg_target *target)
+{
+	assert(is_ody_ocmb_chip(target) == true);
+	// get SBE FFDC info
+	bufPtr_t bufPtr;
+	uint32_t ffdcLen = 0;
+	uint32_t status = 0;
+
+	// sbe_ffdc_get internall uses chipop target so probe it
+	pdbg_target *co = get_ody_chipop_target(target);
+	if (pdbg_target_probe(co) != PDBG_TARGET_ENABLED) {
+		throw sbeError_t(exception::PDBG_TARGET_INVALID);
+	}
+
+	if (sbe_ffdc_get(target, &status, bufPtr.getPtr(), &ffdcLen)) {
+		log(level::ERROR, "POZ  sbe_ffdc_get function failed");
+		throw sbeError_t(exception::SBE_FFDC_GET_FAILED);
+	}
+
+	if (status == (SBEFIFO_PRI_UNKNOWN_ERROR | SBEFIFO_SEC_HW_TIMEOUT)) {
+		log(level::ERROR, "POZ SBE chipop timeout reported(%s)",
+		    pdbg_target_path(target));
+		return sbeError_t(exception::SBE_CMD_TIMEOUT);
+	}
+
+	// Handle empty buffer.
+	if (!ffdcLen) {
+		// log message and return.
+		log(level::ERROR, "POZ empty SBE FFDC returned (%s)",
+		    pdbg_target_path(target));
+		return sbeError_t(exception::SBE_FFDC_NO_DATA);
+	}
+
+	// first store all the data based on slid
+	std::multimap<uint16_t, std::vector<uint8_t>> mulSlidData;
+	uint32_t offset = 0;
+	while ((offset < ffdcLen) &&
+		ffdcLen >= (offset + minFFDCPackageWordSizePOZ)) {
+		pozFfdcHeader *ffdc =
+			reinterpret_cast<pozFfdcHeader *>(bufPtr.getData() + offset);
+		uint16_t magicBytes = ntohs(ffdc->magicByte);
+		assert(magicBytes == pozFfdcMagicCode);
+		uint32_t lenInBytes =
+			ntohs(ffdc->lengthinWords) * sizeof(uint32_t);
+		uint16_t slid = ntohs(ffdc->slid);
+		std::vector<uint8_t> data;
+		data.reserve(lenInBytes);
+		std::copy(bufPtr.getData() + offset,
+			  bufPtr.getData() + offset + lenInBytes,
+			  std::back_inserter(data));
+		mulSlidData.emplace(std::make_pair(slid, data));
+		offset += lenInBytes;
+	}
+	// now combine data with same slid numbers
+	std::map<uint16_t, std::vector<uint8_t>> slidData;
```

Comment:
Unordered multimap if feasible

Reply:
her i do not want duplicate keys so used unordered map